File: TOOLKIT/hdfeos/samples/ReadDimscaleGrid.c
#include "hdf.h"
#include "HdfEosDef.h"

/*
 * In this example we will (1) open the "GridFile" HDF file, (2) attach to
 * the "Grid1" grid, and (3) get dimension scale for a dimension in a filed.
 */


main()
{

    intn            status, i, j;
    int32           gdfid, GDid1, GDid2;
    int32           xdim, ydim;
    float64         *datbuf_flt64;
    int32           *datbuf_int32;
    int32           nbands;
    int32           ntime;
    intn            buffsize;
    int32           data_type;
    int32           xdimsize, ydimsize, dimsize;
    int32           xdimsize1, ydimsize1, dimsize1;
    intn            len;
    char            label[16], unit[16], format[16];
    float64         upleftpt[2], lowrightpt[2];

    /*
     * We first open the HDF grid file, "GridFile.hdf".  Because this file
     * already exist and we wish to write to it, we use the DFACC_RDWR access
     * code in the open statement.  The GDopen routine returns the grid file
     * id, gdfid, which is used to identify the file in subsequent routines.
     */

    gdfid = GDopen("GridFile_created_with_hadeos_sample_file_writer_of_HDFEOS2_version_219_or_higher_release.hdf", DFACC_RDWR);


    /*
     * If the grid file cannot be found, GDopen will return -1 for the file
     * handle (gdfid).  We there check that this is not the case before
     * proceeding with the other routines.
     * 
     * The GDattach routine returns the handle to the existing grid "Grid1",
     * GDid.  If the grid is not found, GDattach returns -1 for the handle.
     */

    if (gdfid != -1)
    {
	/* the field Pollution has Time,YDim,XDim dimensions. 
	   see SetupGrid.c and GDdefineGDflds.c 
	   xdim = 120;
	   ydim = 200;
	   Time dim is set to 10
	*/

	GDid1 = GDattach(gdfid, "UTMGrid");
	if (GDid1 == -1)
	  {
	    printf("\t\tError: Cannot attach to grid \"UTMGrid\"\n");
	    return -1;
	  }

	/* we should get
	   xdim = 120;
	   ntime = 10;
	*/

	/* get XDim */
	status = GDgridinfo(GDid1, &xdimsize, &ydimsize,
			    upleftpt, lowrightpt);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot get grid info for \"UTMGrid\"\n");
	    return -1;
	  }
	printf("X dim size, Y dim size (UTMGrid): %d %d\n", 
	       xdimsize, ydimsize);
	printf("Up left pt (UTMGrid): %lf %lf\n", 
	       upleftpt[0], upleftpt[1]);
	printf("Low right pt (UTMGrid): %lf %lf\n", 
	       lowrightpt[0], lowrightpt[1]);

	buffsize = GDgetdimscale(GDid1, "Pollution", "XDim", &xdimsize1, &data_type, NULL);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for XDim dimemnsion in field \"Pollution\"\n");
	    return -1;
	  }
	else
	  {
	    printf(" \t\tFor Pollution field XDim: xdimsize= %d  buffsize = %d data_type = %d\n",xdimsize, buffsize,data_type);
	    printf(" \t\tFor Pollution field XDim: xdimsize1= %d  buffsize = %d data_type = %d\n",xdimsize1, buffsize,data_type);
	    datbuf_flt64 = (float64 *)malloc(buffsize);
	  }

	buffsize = GDgetdimscale(GDid1, "Pollution", "XDim", &xdimsize1, &data_type, (void *)datbuf_flt64);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for XDim dimemnsion in field \"Pollution\"\n");
	    return -1;
	  }
	else
	  {
	    printf(" \t\tFor Pollution field XDim dimension scale values:\n");
	    for (i=0; i<xdimsize; i++)
	      {
		printf("i = %d       datbuf_flt64 = %lf\n", i, datbuf_flt64[i]);
	      }
	    printf(" \n");
	  }

	if(datbuf_flt64 != NULL)
	  {
	    free(datbuf_flt64);
	    datbuf_flt64 = NULL;
	  }

	/* get str attributes */
	len = 15;
	status = GDgetdimstrs(GDid1, "Pollution", "XDim",
			      label, unit, format, len);
	if(status == 0)
	  {
	    printf(" \t\tFor Pollution field XDim dimension scale attr. strings:\n");
	    printf(" \t\tlabel = %s   unit = %s   format = %s \n\n", label, unit, format);
	  }


	/* get Time */

	dimsize = GDdiminfo(GDid1, "Time");
	buffsize = GDgetdimscale(GDid1, "Pollution", "Time", &dimsize1, &data_type, NULL);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for Time dimemnsion in field \"Pollution\"\n");
	    return -1;
	  }
	else
	  {
	    printf(" \t\tFor Pollution field Time: dimsize= %d buffsize = %d data_type = %d\n",dimsize, buffsize,data_type);
	    printf(" \t\tFor Pollution field Time: dimsize1= %d buffsize = %d data_type = %d\n",dimsize1, buffsize,data_type);
	    datbuf_int32 = (int32 *)malloc(buffsize);
	  }

	buffsize = GDgetdimscale(GDid1, "Pollution", "Time", &dimsize1, &data_type, (void *)datbuf_int32);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for Time dimemnsion in field \"Pollution\"\n");
	    return -1;
	  }
	else
	  {
	    printf(" \t\tFor Pollution field Time dimension scale values:\n");
	    for (i=0; i<dimsize; i++)
	      {
		printf("i = %d       datbuf_int32 = %d\n", i, datbuf_int32[i]);
	      }
	    printf(" \n");
	  }

	/* get str attributes */
	len = 15;
	status = GDgetdimstrs(GDid1, "Pollution", "Time",
			      label, unit, format, len);
	if(status == 0)
	  {
	    printf(" \t\tFor Pollution field Time dimension scale attr. strings:\n");
	    printf(" \t\tlabel = %s   unit = %s   format = %s \n\n", label, unit, format);
	  }

	if(datbuf_int32 != NULL)
	  {
	    free(datbuf_int32);
	    datbuf_int32 = NULL;
	  }

	  /* the field Spectra has Bands,YDim,XDim dimensions. 
	   see SetupGrid.c and GDdefineGDflds.c 
	   xdim = 100;
	   ydim = 100;
	   Bands dim is set to 3
	  */

	GDid2 = GDattach(gdfid, "PolarGrid");
	if (GDid2 == -1)
	  {
	    printf("\t\tError: Cannot attach to grid \"PolarGrid\"\n");
	    return -1;
	  }

	/* we should get
	   ydim = 100;
	   nbands = 3;
	*/

	/* get YDim */
	status = GDgridinfo(GDid2, &xdimsize, &ydimsize,
			    upleftpt, lowrightpt);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot get grid info for \"PolarGrid\"\n");
	    return -1;
	  }
	printf("X dim size, Y dim size (PolarGrid): %d %d\n", 
	       xdimsize, ydimsize);
	printf("Up left pt (UTMGrid): %lf %lf\n", 
	       upleftpt[0], upleftpt[1]);
	printf("Low right pt (UTMGrid): %lf %lf\n", 
	       lowrightpt[0], lowrightpt[1]);


	buffsize = GDgetdimscale(GDid2, "Spectra", "YDim", &ydimsize1, &data_type, NULL);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for YDim dimemnsion in field \"Spectra\"\n");
	    return -1;
	  }
	else
	  {
	    printf(" \t\tFor Spectra field YDim: ydimsize = %d buffsize = %d data_type = %d\n",ydimsize, buffsize,data_type);
	    printf(" \t\tFor Spectra field YDim: ydimsize1 = %d buffsize = %d data_type = %d\n",ydimsize1, buffsize,data_type);
	    datbuf_flt64 = (float64 *)malloc(buffsize);
	  }

	   buffsize = GDgetdimscale(GDid2, "Spectra", "YDim", &ydimsize1, &data_type, (void *)datbuf_flt64);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for YDim dimemnsion in field \"Spectra\"\n");
	    return -1;
	  }
	else
	  {
	    printf(" \t\tFor Spectra field YDim dimension scale values:\n");
	    for (i=0; i<ydimsize; i++)
	      {
		printf("i = %d       datbuf_flt64 = %lf\n", i, datbuf_flt64[i]);
	      }
	    printf(" \n");
	  }

	/* get str attributes */
	len = 15;
	status = GDgetdimstrs(GDid2, "Spectra", "YDim",
			      label, unit, format, len);
	if(status == 0)
	  {
	    printf(" \t\tFor Spectra field YDim dimension scale attr. strings:\n");
	    printf(" \t\tlabel = %s   unit = %s   format = %s \n\n", label, unit, format);
	  }

	if(datbuf_flt64 != NULL)
	  {
	    free(datbuf_flt64);
	    datbuf_flt64 = NULL;
	  }

	/* get Bands */

	dimsize = GDdiminfo(GDid2, "Bands");
	buffsize = GDgetdimscale(GDid2, "Spectra", "Bands", &dimsize1, &data_type, NULL);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for Bands dimemnsion in field \"Spectra\"\n");
	    return -1;
	  }
     	else
	  {
	    printf(" \t\tFor Spectra field Bands: dimsize = %d  buffsize = %d data_type = %d\n",dimsize, buffsize,data_type);
	    printf(" \t\tFor Spectra field Bands: dimsize1 = %d  buffsize = %d data_type = %d\n",dimsize1, buffsize,data_type);
	    datbuf_int32 = (int32 *)malloc(buffsize);
	  }

	buffsize = GDgetdimscale(GDid2, "Spectra", "Bands", &dimsize1, &data_type, (void *)datbuf_int32);
	if (buffsize == -1)
	  {
	    printf("\t\tError: Cannot get Dimension Scale for Bands dimemnsion in field \"Spectra\"\n");
	    return -1;
	  }
     	else
	  {
	    printf(" \t\tFor Spectra field Bands dimension scale values:\n");
	    for (i=0; i<dimsize; i++)
	      {
		printf("i = %d       datbuf_int32 = %d\n", i, datbuf_int32[i]);
	      }
	    printf(" \n");
	  }
	
	/* get str attributes */
	len = 15;
	status = GDgetdimstrs(GDid2, "Spectra", "Bands",
			      label, unit, format, len);
	if(status == 0)
	  {
	    printf(" \t\tFor Spectra field Bands dimension scale attr. strings:\n");
	    printf(" \t\tlabel = %s   unit = %s   format = %s \n\n", label, unit, format);
	  }

	if(datbuf_flt64 != NULL)
	  {
	    free(datbuf_flt64);
	    datbuf_flt64 = NULL;
	  }

	if(datbuf_int32 != NULL)
	  {
	    free(datbuf_int32);
	    datbuf_int32 = NULL;
	  }
    }

    GDdetach(GDid1);
    GDdetach(GDid2);

    GDclose(gdfid);

    return 0;
}

File: TOOLKIT/hdfeos/samples/readdimscalegrid.f
!/*
! * In this example we will (1) open the "GridFile" HDF file, (2) attach to
! * the "gdid" grid, and (3) set dimension scale for a dimension in a filed.
! */

	program readdimscalegrid
	
	integer       j, status, gddetach, gdclose
	integer*4     gdfid, gdid1, gdid2, gdopen, gdattach
	integer*4     gdgetdimscale, gdgetdimstrs
	real*8        veg1(120),veg2(200)
	real*8        temp1(100),temp2(100)
	integer*4     bands(3)
	integer*4     xdimsize, ydimsize,dimsize
	integer*4     xdimsize1, ydimsize1,dimsize1
	real*8        upleftpt(2), lowrightpt(2)
	integer*4     ntype(32)
	character*16  label, unit, format
	integer*4     len
	integer*4     nbands
	integer DFACC_RDWR
	parameter (DFACC_RDWR=3)
	integer DFNT_FLOAT32
	parameter (DFNT_FLOAT32=5)
	integer DFNT_FLOAT64
	parameter (DFNT_FLOAT64=6)
	integer DFNT_INT32
	parameter (DFNT_INT32=24)

!    /*
!     * We first open the HDF grid file, "GridFile.hdf".  Because this file
!     * already exist and we wish to write to it, we use the DFACC_RDWR access
!     * code in the open statement.  The gdopen routine returns the grid file
!     * id, gdfid, which is used to identify the file in subsequent routines.
!     */

	gdfid = gdopen("GridFile_created_with_hadeos_sample_file_write"//
     1"r_of_HDFEOS2_version_219_or_higher_release.hdf", DFACC_RDWR)

!    /*
!     * If the grid file cannot be found, gdopen will return -1 for the file
!     * handle (gdfid).  We there check that this is not the case before
!     * proceeding with the other routines.
!     * 
!     * The GDattach routine returns the handle to the existing grid "gdid",
!     * gdid.  If the grid is not found, gdattach returns -1 for the handle.
!     */	
	if (gdfid .ne. -1) then
	   
	   gdid1 = gdattach(gdfid, "UTMGrid")
	   
	   if (gdid1 .ne. -1) then
	      
	      status = gdgridinfo(gdid1, xdimsize, ydimsize,
     1	      upleftpt, lowrightpt)
	      
	      write(*,*) 'X dim size, Y dim size (PolarGrid): ', 
     1	      xdimsize, ydimsize
	      write(*,*) 'Up left pt (UTMGrid): ', 
     1	      upleftpt(1), upleftpt(2)
	      write(*,*) 'Low right pt (UTMGrid): ', 
     1	      lowrightpt(1), lowrightpt(2)
	      
	      buffsize = gdgetdimscale(gdid1, "Vegetation",
     1         "XDim", xdimsize1, ntype,veg1)
	      if (buffsize .eq. -1) then	   
		 write(*,*) 'Error:Cannot get Dim Scale for XDim',
     1                      ' in field Vegetation'
		 stop  
	      else
		 write(*,*) 'For Vegetation field XDim: xdimsize1 = ', 
     1                      xdimsize1
		 write(*,*) 'For Vegetation field XDim: buffsize = ', 
     1                      buffsize
	      endif
	      
	      write(*,*) 'First 10 elements for Vegetation field XDim',
     1                   ' dimension scale values:'
	      do j=1,10
		 write(*,*) 'j =',j,'         veg1 = ',veg1(j)
	      enddo
	      write(*,*) '  '
	      
	      buffsize = gdgetdimscale(gdid1, "Vegetation", "YDim",
     1                                  ydimsize1,ntype,veg2)
	      if (buffsize .eq. -1) then	   
		 write(*,*) 'Error:Cannot get Dim Scale for YDim',
     1                      ' in field Vegetation"'
		 stop  
	      else
		 write(*,*) 'For Vegetation field YDim: ydimsize1 = ', 
     1                      ydimsize1
		 write(*,*) 'For Vegetation field YDim: buffsize = ', 
     1                      buffsize
	      endif
	      
	      write(*,*) 'First 10 elements for Vegetation field YDim',
     1                   ' dimension scale values:'
	      do j=1,10
		 write(*,*) 'j =',j,'         veg2 = ',veg2(j)
	      enddo
	      write(*,*) '  '
	      label = ''
	      unit = ''
	      format = ''
	      len = 15
	      status = gdgetdimstrs(gdid1, "Vegetation", "YDim", 
     1	      label, unit, format, len)
	      
	      
	      if(status .eq. 0) then	  
		 write(*,*) 'For Vegetation field YDim',
     1                      ' dimension scale attr. strings:'
		 write(*,*) 'label =',label,'  unit =',
     1                      unit,'  format = ',format
	      endif
	      write(*,*) ' '
	      status = gddetach(gdid1)
	   endif

	   gdid2 = gdattach(gdfid, "PolarGrid")
	   if (gdid2 .ne. -1) then
	      
!	/* we should get
!       ydim = 100	
!       nbands = 3;	
!	*/		
!        /* get YDim */	
	   
	      status = gdgridinfo(gdid2, xdimsize, ydimsize,
     1	      upleftpt, lowrightpt)
	      
	      write(*,*) 'X dim size, Y dim size (PolarGrid): ', 
     1	      xdimsize, ydimsize
	      write(*,*) 'Up left pt (PolarGrid): ', 
     1	      upleftpt(1), upleftpt(2)
	      write(*,*) 'Low right pt (PolarGrid): ', 
     1	      lowrightpt(1), lowrightpt(2)
	      
	      buffsize = gdgetdimscale(gdid2, "Spectra", "YDim",
     1	                               ydimsize1,ntype, temp2)
	      if (buffsize .eq. -1) then	   
		 write(*,*) 'Error:Cannot get Dim Scale for YDim',
     1                      ' in field Spectra'
		 stop  
	      else
		 write(*,*) 'For Spectra field YDim: ydimsize1 = ', 
     1                      ydimsize1
		 write(*,*) 'For Spectra field YDim: buffsize = ', 
     1                       buffsize
	      endif
	      
	      write(*,*) 'First 10 elements for Spectra field YDim',
     1                   ' dimension scale values:'
	      do j=1,10
		 write(*,*) 'j =',j,'         temp2 = ',temp2(j)
	      enddo
	      write(*,*) '  '
	      
!	/* get str attributes */
	      len = 15
	      status = gdgetdimstrs(gdid2, "Spectra", "YDim", 
     1	      label, unit, format, len)
	      
	      
	      if(status .eq. 0) then	  
		 write(*,*) 'For Spectra field YDim',
     1                      ' dimension scale attr. strings:'
		 write(*,*) 'label =',label,'  unit =',
     1                      unit,'  format = ',format
	      endif
	      write(*,*) ' '
	      
	      dimsize = gddiminfo(gdid2, "Bands");
	      status = gdgetdimscale(gdid2, "Spectra", "Bands", 
     1                               nbands,ntype, bands)
	      write(*,*) 'For Spectra field Bands: nbands = ', 
     1	                  nbands
	      write(*,*) 'For Spectra field Bands',
     1	                 'dimension scale values:'

	      do j=1,nbands
		 write(*,*) 'j =',j,'         Bnads = ',bands(j)
	      enddo
	      write(*,*) '  '
	      
	      status = gdgetdimstrs(gdid2, "Spectra", "Bands", 
     1	      label, unit, format, len)
	      
	      if(status .eq. 0) then
		 write(*,*) 'For Spectra field Bands',
     1                      ' dimension scale attr. strings:'
		 write(*,*) 'label =',label,'  unit =',
     1                      unit,'  format = ',format
	      endif
	      
	      status = gddetach(gdid2)
	   endif
	   status = gdclose(gdfid)
	endif
	stop
	end
